pyccl/halos/massdef.py

```diff
@@ -0,0 +1,247 @@
+    found by solving the equation:
+        f(c_old) * D_old = f(c_new) * D_new
+    where f(x) = x^3/[ln(1+x) - x/(1+x)].
+
+    Args:
+        cosmo (:obj:`Cosmology`): A Cosmology object.
+        c_old (float or array_like): concentration to translate from.
+        Delta_old (float): Delta parameter associated to the input
+            concentration. See description of the MassDef class.
+        Delta_new (float): Delta parameter associated to the output
+            concentration.
+
+    Returns:
+        float or array_like: concentration parameter for the new
+        mass definition.
+    """
+    status = 0
+    c_old_use = np.atleast_1d(c_old)
+    c_new, status = lib.convert_concentration_vec(cosmo.cosmo,
+                                                  Delta_old, c_old_use,
+                                                  Delta_new, c_old_use.size,
+                                                  status)
+    if np.isscalar(c_old):
+        c_new = c_new[0]
+
+    check(status)
+    return c_new
+
+
+class MassDef(object):
+    """Halo mass definition. Halo masses are defined in terms of an overdensity
+    parameter Delta and an associated density X (either the matter density or
+    the critical density):
+        M = 4 * pi * Delta * rho_X * R^3 / 3
+    where R is the halo radius. This object also holds methods to translate
+    between R and M, and to translate masses between different definitions
+    if a concentration-mass relation is provided.
+
+    Args:
+        Delta (float): overdensity parameter. Pass 'vir' if using virial
+            overdensity.
+        rho_type (string): either 'critical' or 'matter'.
+        c_m_relation (function, optional): concentration-mass relation.
+            Provided as a `Concentration` object, or a string corresponding
+            to one of the supported concentration-mass relations.
+            If `None`, no c(M) relation will be attached to this mass
+            definition (and hence one can't translate into other definitions).
+    """
+    def __init__(self, Delta, rho_type, c_m_relation=None):
+        # Check it makes sense
+        if (Delta != 'fof') and (Delta != 'vir'):
+            if Delta <= 0:
+                raise ValueError("Delta must be a positive number")
+        self.Delta = Delta
+        # Can only be matter or critical
+        if rho_type not in ['matter', 'critical']:
+            raise ValueError("rho_type must be either \'matter\' "
+                             "or \'critical\'")
+        self.rho_type = rho_type
+        self.species = species_types[rho_type]
+        # c(M) relation
+        if c_m_relation is None:
+            self.concentration = None
+        else:
+            self._concentration_init(c_m_relation)
+
+    def __eq__(self, other):
+        """ Allows you to compare two mass definitions
+        """
+        return (self.Delta == other.Delta) and \
```

Comment:
Here we are doing exact equality again on delta. See discussion above for how to potentially handle this.

Reply:
I think this is fine. We really want to check for exactly equal Deltas in this case (we want to see if two mass definitions are actually the same one, so I would enforce them being actually the same). If the Deltas differ by a bit, then there will be a mass conversion, but this shouldn't matter for practical purposes.